### PR TITLE
Signature payload format v1

### PR DESCRIPTION
**Pull request by Geal**

### CODE REVIEW

File: SPECIFICATIONS.md [hunk 1]
```diff
@@ -780,13 +782,52 @@ corresponding to the last public key, to sign a new block and attenuate the
...
+The data covered by the signature algorithm depends on the `version` field of
+the `SignedBlock` message. If the field is absent, it defaults to version 0.
+
+##### Version 0 (deprecated)
+
+To sign the block at index `n`, we would have:
+- `data_n`: the serialized Datalog
+- `pk_n+1`: the next public key
+- `alg_n+1`: the little endian representation of the signature algorithm for `pk_n+1`
+- `external_sig_n`: te optional external signature of the block
```
Comment (divarvel) [inline SPECIFICATIONS.md:796]:
```suggestion
- `external_sig_n`: the optional external signature of the block
```

File: schema.proto [hunk 2]
```diff
@@ -99,13 +100,35 @@ message TermV2 {
     bool bool = 6;
     TermSet set = 7;
     Empty null = 8;
+    Array array = 9;
```
Comment (divarvel) [inline schema.proto:103]: this should already be part of the `dev` branch, maybe you have started your work on an outdated version?

Comment (Geal) [reply schema.proto]: yup, this was fixed by merging dev in here

**Comment by Geal** [comment]
updated with the lastest changes from [a1a2d59](https://github.com/biscuit-auth/biscuit-rust/pull/230/commits/a1a2d59f2edb5a48bb374bff3b6ed8b2aa38c2a6). This should work well now. There's one last decision here: do we update the sealing signature payload or remove sealing entirely?

**Comment by divarvel** [comment]
This looks really good.

A couple suggestions:

- separate the authority block signature description from subsequent blocks, this will prevent invalid combinations such as having an external signature defined, but no previous signature (this does not change anything to the actual behaviour, it just makes the spec clearer). This would also help describing how to sign attenuation blocks as `block n+1`, with n >= 0, instead of having a single description for `block n` with n >= 0 with references to `n-1` (which is not defined for n = 0)
- reorder signature elements to put the optional elements at the end
- prefix the external signature input with `\0EXTERNAL\0` to avoid having the same input for the regular and external signatures

File: SPECIFICATIONS.md [hunk 3]
```diff
@@ -787,13 +789,108 @@ corresponding to the last public key, to sign a new block and attenuate the
...
+the `SignedBlock` message. If the field is absent, it defaults to version 0.
+
+##### Version 0 (deprecated)
+
+This defines the block signature payload v0.
+
+To sign the authority block, we have:
+- `data_0`: the serialized Datalog
+- `pk_1`: the next public key
+- `alg_1`: the little endian representation of the signature algorithm for `pk_n+1`
```
Comment (divarvel) [inline SPECIFICATIONS.md:804]:
```suggestion
- `alg_1`: the little endian representation of the signature algorithm for `pk_1`
```

File: SPECIFICATIONS.md [hunk 4]
```diff
@@ -787,13 +789,108 @@ corresponding to the last public key, to sign a new block and attenuate the
...
+
+The authority block signature payload v0 would be the concatenation of:
+- `data_0`
+- `pk_1`
+- `alg_1`
+
+To sign the block at index `n+1`, we have:
+- `data_n+1`: the serialized Datalog
+- `pk_n+2`: the next public key
+- `alg_n+2`: the little endian representation of the signature algorithm for `pk_n+1`
```
Comment (divarvel) [inline SPECIFICATIONS.md:814]:
```suggestion
- `alg_n+2`: the little endian representation of the signature algorithm for `pk_n+2`
```

File: SPECIFICATIONS.md [hunk 5]
```diff
@@ -787,13 +789,108 @@ corresponding to the last public key, to sign a new block and attenuate the
...
+This format is not supported anymore and should be replaced by version 1.
+
+##### Version 1
+
+This defines the block signature payload v1.
+
+To sign the authority block, we have:
+- `data_0`: the serialized Datalog
+- `pk_1`: the next public key
+- `alg_1`: the little endian representation of the signature algorithm for `pk_n+1`
```
Comment (divarvel) [inline SPECIFICATIONS.md:844]:
```suggestion
- `alg_1`: the little endian representation of the signature algorithm for `pk_1`
```

File: SPECIFICATIONS.md [hunk 6]
```diff
@@ -852,8 +951,8 @@ Token {
 Blocks generated by a trusted third party can carry an *extra* signature to provide a proof of their
 origin. Same as regular signatures, they rely on Ed25519.
 
-The external signature for block `n+1`, with `(external_pk, external_sk)` is `external_sig_n+1 = sign(external_sk, data_n+1 + alg_n+1 + pk_n+1)`.
-It's quite similar to the regular signature, with a crucial difference: the public key appended to the block payload is the one _carried_ by block `n` (and which is used to verify block `n+1`).
+The external signature for block `n+1`, with `(external_pk, external_sk)` is `external_sig_n+1`, the signature of the payload in format "external signature payload v1" by `external_sk`.
+It is quite similar to the regular signature, with a crucial difference: the public key appended to the block payload is the one _carried_ by block `n` (and which is used to verify block `n+1`).
```
Comment (divarvel) [inline SPECIFICATIONS.md:955]: i think we can remove this line, or maybe move it to the description of the v0 sig algorithm, as it does not really apply to v1

File: SPECIFICATIONS.md [hunk 7]
```diff
@@ -787,13 +789,88 @@ corresponding to the last public key, to sign a new block and attenuate the
...
+- the binary representation of the ASCII string "\0NEXTKEY\0"
+- `pk_1`: the next public key
+
+To sign the block at index `n+1`, the signed payload format is the concatenation of:
+- the binary representation of the ASCII string "\0BLOCK\0"
+- the binary representation of the ASCII string "\0VERSION\0"
+- the little endian representation of the version of the signature payload format
+- the binary representation of the ASCII string "\0PAYLOAD\0"
+- `data_n+1`: the serialized Datalog
+- the binary representation of the ASCII string "\0PREVSIG\0"
```
Comment (divarvel) [inline SPECIFICATIONS.md:847]: we could put prevsig after algorithm/nextkey to have a bit more consistency between the authority and attenuation blocks, and group signatures. not a big deal, probably just me being nitpicky

Comment (Geal) [reply SPECIFICATIONS.md:847]: done in [5d99d3d](https://github.com/biscuit-auth/biscuit/pull/175/commits/5d99d3d6afb2a019a0732a28a4ff97a442252aa5)

**Review by divarvel — APPROVED** [review]

**Pull request merged**